private/pkg/manifest/manifest.go

```diff
@@ -0,0 +1,217 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Manifests are a list of paths and their hash digests, canonically ordered by
+// path in increasing lexographical order. Manifests are encoded as:
+//
+//	<hash type>:<digest>[SP][SP]<path>[LF]
+//
+// "shake256" is the only supported hash type. The digest is 64 bytes of hex
+// encoded output of SHAKE256. See golang.org/x/crypto/sha3 and FIPS 202 for
+// details on the SHAKE hash.
+package manifest
+
+import (
+	"bufio"
+	"bytes"
+	"encoding"
+	"errors"
+	"fmt"
+	"io"
+	"sort"
+	"strings"
+)
+
+var errNoFinalNewline = errors.New("partial record: missing newline")
+
+// Error occurs when a manifest is malformed.
+type Error struct {
```

Comment:
None of this is exposed, aside from in the error message. Are there plans to add accessors? If not, do we need an exported type for this? You could also construct this using `fmt.Errorf` with `%w` directive to wrap an underlying error and add the line number and message.

Reply:
Fixed.